### PR TITLE
Resident Directory Filter

### PR DESCRIPTION
**Pull request by phamkelly17**
## Notion task link

[Resident Directory Filter](https://www.notion.so/uwblueprintexecs/Resident-Directory-Filters-90cb564cd66345bfbd14221bda03c0a5)

## Implementation description
* Implement backend for filtering residents
* can filter by `resident_id`, `building_id`, `date_range`
* see Notion documentation for get_residents()

## Steps to test
1. Test endpoints on Postman
2. Test individually all filters
3. Test together

## What should reviewers focus on?
* 

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR
- [x] If I have made API changes, I have updated the [REST API Docs](https://www.notion.so/uwblueprintexecs/REST-Endpoints-05ce60312bb943439dfda42bb1318536)
- [x] IF I have made changes to the db/models, I have updated the [Data Models Page](https://www.notion.so/uwblueprintexecs/Data-Models-760f8aa06b244eb0842c079ad77987b0)
- [x] I have documented this PR in the [Production Release Page](https://www.notion.so/uwblueprintexecs/fe28a97bfa5648d3bc3795b32b694acd?v=5565cfb35dcc45bb84f6528cb09517cd)
- [x] I have updated other Docs as needed

### CODE REVIEW

**Comment by braydonwang** [comment]
Just wondering, is this the right endpoint url to call when we want to filter by date range but only provide a start date?

```
localhost:8080/residents?filters={"date_range":{"date_joined": "2023-11-23", "date_left": None}}
```

**Comment by phamkelly17** [comment]
@braydonwang this is the correct endpoint:


`http://localhost:8080/residents?filters={"date_range": ["2024-1-31", null]}`

File: backend/app/services/implementations/residents_service.py [hunk 1]
```diff
@@ -120,6 +120,50 @@ def get_residents(
...
+                    if date_joined is not None:
+                        date_joined = datetime.strptime(
+                            date_joined, "%Y-%m-%d"
+                        ).replace(hour=0, minute=0)
+                    if date_left is not None:
+                        date_left = datetime.strptime(date_left, "%Y-%m-%d").replace(
+                            hour=0, minute=0
+                        )
+
+                residents_results = Residents.query.join(
```
Comment (owen-sellner) [inline backend/app/services/implementations/residents_service.py:153]: This probably can be moved out of the if-else statements since it is always used.

**Review by owen-sellner — APPROVED** [review]
LGTM!!! I added the one nit that we already talked about.

**Review by connor-bechthold — APPROVED** [review]

**Pull request merged**